File: mobile/ios/WaldoKit/Sources/WaldoKit/Screens/Groups/GroupDetailScreen.swift
import SwiftUI

/// specs/004-ios-client.md §3.4/§3.5 (001 §12.3–12.9; 005 §2.3; specs/007-public-join-links.md) —
/// composes ONLY design-system components (bar `ShareLink`/`DatePicker`/`.confirmationDialog`, the
/// same documented system-primitive-exception pattern used by the invite/history screens; Android's
/// equivalent exception is Material3 `AlertDialog`, specs/003 §4.3/§12.2). Rendering follows the
/// 005 §2.3 lazy-enforcement matrix: `active` shows the map entry point; roster is hidden for
/// non-owner members during `ended` (grace); rename/extend/rotate are owner-and-active/ended-only
/// (archived groups reject PATCH, 001 §12.4); kick/leave/delete stay available in every
/// non-expired state, matching the matrix exactly. `GROUP_EXPIRED` renders as a persistent notice
/// (`.expired`, same pattern as `GroupMapScreen`) requiring an explicit "Back to groups" tap, not a
/// silent exit. Since 007/I6, the share card's `ShareLink` and on-device QR carry the canonical
/// `https://{joinLinkHost}/g#CODE` link rather than plain text.
public struct GroupDetailScreen: View {
    @Environment(\.theme) private var theme
    @ObservedObject private var viewModel: GroupDetailViewModel
    @State private var editedName: String = ""
    @State private var extendedEndsAt: Date = Date().addingTimeInterval(24 * 3600)
    /// Every owner mutation the spec (003 §12.2/004 §3.4) requires a confirm step for
    /// (rename/extend/end-now/rotate/kick/delete) funnels through this single pending-action +
    /// `.confirmationDialog`, rather than one boolean flag per action.
    @State private var pendingConfirmation: PendingConfirmation?
    /// specs/007-public-join-links.md §1, specs/004-ios-client.md §3.5 — the deployment constant
    /// the share link/QR are built against (`AppConfig.joinLinkHost`).
    private let joinLinkHost: String
    private let onSelectMap: () -> Void
    private let onExit: () -> Void

    public init(
        viewModel: GroupDetailViewModel,
        joinLinkHost: String = AppConfig.defaultJoinLinkHost,
        onSelectMap: @escaping () -> Void,
        onExit: @escaping () -> Void
    ) {
        self.viewModel = viewModel
        self.joinLinkHost = joinLinkHost
        self.onSelectMap = onSelectMap
        self.onExit = onExit
    }

    public var body: some View {
        VStack(spacing: 0) {
            WaldoNavBar("Group")
            content
        }
        .background(theme.colors.surfaceVariant)
        .task { await viewModel.load() }
        .onChange(of: viewModel.exitReason) { reason in
            if reason != nil { onExit() }
        }
        .confirmationDialog(
            "Are you sure?", isPresented: isConfirmationPresented,
            titleVisibility: .visible, presenting: pendingConfirmation
        ) { pending in
            Button(pending.actionTitle, role: .destructive) {
                Task { await perform(pending) }
            }
            Button("Cancel", role: .cancel) {}
        } message: { pending in
            Text(pending.message)
        }
    }

    private var isConfirmationPresented: Binding<Bool> {
        Binding(get: { pendingConfirmation != nil }, set: { if !$0 { pendingConfirmation = nil } })
    }

    private func perform(_ pending: PendingConfirmation) async {
        switch pending {
        case .rename: await viewModel.rename(name: editedName)
        case .extend: await viewModel.extend(endsAt: extendedEndsAt)
        case .endNow: await viewModel.endNow()
        case .rotateCode: await viewModel.rotateCode()
        case .kick(let userId, _): await viewModel.kick(userId: userId)
        case .delete: await viewModel.deleteGroup()
        }
    }

    @ViewBuilder
    private var content: some View {
        switch viewModel.state {
        case .loading:
            LoadingStateView(message: "Loading group…")
        case .error(let message):
            ErrorStateView(message: message) {
                Task { await viewModel.load() }
            }
        case .expired:
            // 005 §2.3 / specs/004 §3.4 — the group ended while this screen was open; a
            // persistent notice the caller must acknowledge, exactly like GroupMapScreen.
            ErrorStateView(message: "This group has ended.", retryTitle: "Back to groups", onRetry: onExit)
        case .loaded(let detail):
            detailView(detail)
        }
    }

    private func detailView(_ detail: GroupDetail) -> some View {
        ScrollView {
            VStack(alignment: .leading, spacing: theme.spacing.md) {
                header(detail)
                if let lastActionError = viewModel.lastActionError {
                    ErrorStateView(message: lastActionError)
                }
                if detail.state == "active" {
                    WaldoButton("Group map") { onSelectMap() }
                }
                if let code = detail.code {
                    shareCodeCard(code: code, name: detail.name)
                }
                rosterOrGraceNotice(detail)
                controls(detail)
            }
            .padding(theme.spacing.md)
        }
        .onAppear { editedName = detail.name }
    }

    private func header(_ detail: GroupDetail) -> some View {
        HStack(alignment: .top) {
            VStack(alignment: .leading, spacing: theme.spacing.xs) {
                Text(detail.name)
                    .font(theme.typography.titleLarge)
                    .foregroundColor(theme.colors.onSurface)
                Text("\(detail.memberCount) member\(detail.memberCount == 1 ? "" : "s")")
                    .font(theme.typography.bodyMedium)
                    .foregroundColor(theme.colors.onSurface.opacity(0.7))
                Text(GroupCountdown.text(from: detail.endsAt))
                    .font(theme.typography.labelSmall)
                    .foregroundColor(theme.colors.onSurface.opacity(0.7))
            }
            Spacer()
            StatusChip(GroupStateChip.label(for: detail.state), kind: GroupStateChip.kind(for: detail.state))
        }
    }

    /// specs/007-public-join-links.md §1/§4, specs/004-ios-client.md §3.5 — the https link is now
    /// the canonical thing shared and encoded as a QR ("the https form is the canonical one for
    /// sharing and QR"); the `waldo://` deep link stays supported for INCOMING opens (§4) but is no
    /// longer what this screen hands out. The QR renders entirely on-device (`GroupJoinQRCodeView`
    /// — CoreImage, never a networked service).
    private func shareCodeCard(code: String, name: String) -> some View {
        let joinLink = GroupDetailViewModel.joinLink(for: code, joinLinkHost: joinLinkHost)
        return WaldoCard {
            VStack(spacing: theme.spacing.sm) {
                Text(GroupDetailViewModel.shareText(for: code, groupName: name))
                    .font(theme.typography.titleMedium)
                    .foregroundColor(theme.colors.onSurface)
                    .multilineTextAlignment(.center)
                GroupJoinQRCodeView(text: joinLink.absoluteString)
                ShareLink(item: joinLink)
            }
        }
    }

    /// 005 §2.3: during `ended` (grace), non-owner members get no roster — only the owner sees it,
    /// to decide on reactivation. `archived` and `active` always carry the full roster.
    @ViewBuilder
    private func rosterOrGraceNotice(_ detail: GroupDetail) -> some View {
        if let members = detail.members {
            VStack(alignment: .leading, spacing: theme.spacing.sm) {
                ForEach(members, id: \.userId) { member in
                    WaldoListRow(title: member.displayName, subtitle: member.role.capitalized) {
                        if viewModel.isOwner && member.role != "owner" {
                            WaldoButton("Kick", style: .secondary) {
                                pendingConfirmation = .kick(userId: member.userId, displayName: member.displayName)
                            }
                        }
                    }
                }
            }
        } else {
            EmptyStateView(title: "This group has ended", message: "The owner can still reactivate it before it's deleted.")
        }
    }

    @ViewBuilder
    private func controls(_ detail: GroupDetail) -> some View {
        if viewModel.isOwner {
            ownerControls(detail)
        } else {
            WaldoButton("Leave group", style: .secondary) {
                Task { await viewModel.leave() }
            }
        }
    }

    /// PATCH-shaped controls (rename/extend/end-now) are hidden once `archived` — the server
    /// rejects PATCH there (`410 GROUP_EXPIRED`, 001 §12.4). Rotate only makes sense while `active`
    /// (the code row is gone otherwise, 005 §2.3). Kick/delete stay available in every state.
    /// Every action here goes through `pendingConfirmation` — none fire on tap alone (003 §12.2,
    /// mirrored by 004 §3.4: "owner controls behind confirm dialogs").
    @ViewBuilder
    private func ownerControls(_ detail: GroupDetail) -> some View {
        VStack(spacing: theme.spacing.sm) {
            if detail.state != "archived" {
                WaldoTextField("Group name", text: $editedName)
                WaldoButton("Rename", style: .secondary) {
                    pendingConfirmation = .rename(newName: editedName)
                }
                DatePicker("New end date", selection: $extendedEndsAt, displayedComponents: [.date, .hourAndMinute])
                    .tint(theme.colors.primary)
                WaldoButton(detail.state == "ended" ? "Reactivate" : "Extend", style: .secondary) {
                    pendingConfirmation = .extend
                }
                WaldoButton("End group now", style: .secondary) {
                    pendingConfirmation = .endNow
                }
            }
            if detail.state == "active" {
                WaldoButton("Rotate code", style: .secondary) {
                    pendingConfirmation = .rotateCode
                }
            }
            WaldoButton("Delete group", style: .secondary) {
                pendingConfirmation = .delete
            }
        }
    }
}

/// specs/003 §12.2 (mirrored by 004 §3.4) — "owner controls behind confirm dialogs: rename,
/// extend/end…, rotate code, kick member, delete group." One case per confirmable owner action;
/// `Identifiable` so `.confirmationDialog(presenting:)` can drive off it directly.
private enum PendingConfirmation: Identifiable, Equatable {
    case rename(newName: String)
    case extend
    case endNow
    case rotateCode
    case kick(userId: String, displayName: String)
    case delete

    var id: String {
        switch self {
        case .rename: return "rename"
        case .extend: return "extend"
        case .endNow: return "endNow"
        case .rotateCode: return "rotateCode"
        case .kick(let userId, _): return "kick-\(userId)"
        case .delete: return "delete"
        }
    }

    var actionTitle: String {
        switch self {
        case .rename: return "Rename"
        case .extend: return "Confirm"
        case .endNow: return "End group"
        case .rotateCode: return "Rotate"
        case .kick: return "Remove"
        case .delete: return "Delete"
        }
    }

    var message: String {
        switch self {
        case .rename(let newName): return "Rename this group to \"\(newName)\"?"
        case .extend: return "Change this group's end date?"
        case .endNow: return "End this group now? Members will lose access immediately."
        case .rotateCode: return "Rotate the join code? The current code will stop working immediately."
        case .kick(_, let displayName): return "Remove \(displayName) from this group?"
        case .delete: return "Delete this group? This can't be undone — everything about it disappears immediately."
        }
    }
}

File: mobile/android/app/src/main/java/com/whereswaldo/android/ui/groups/GroupDetailScreen.kt
package com.whereswaldo.android.ui.groups

import android.content.Intent
import android.widget.Toast
import androidx.compose.foundation.Image
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.RowScope
import androidx.compose.foundation.layout.aspectRatio
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.FilterQuality
import androidx.compose.ui.graphics.asImageBitmap
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.tooling.preview.Preview
import com.whereswaldo.android.ui.designsystem.WaldoTheme
import com.whereswaldo.android.ui.designsystem.components.WaldoButton
import com.whereswaldo.android.ui.designsystem.components.WaldoButtonStyle
import com.whereswaldo.android.ui.designsystem.components.WaldoCard
import com.whereswaldo.android.ui.designsystem.components.WaldoErrorState
import com.whereswaldo.android.ui.designsystem.components.WaldoListRow
import com.whereswaldo.android.ui.designsystem.components.WaldoLoadingState
import com.whereswaldo.android.ui.designsystem.components.WaldoSectionHeader
import com.whereswaldo.android.ui.designsystem.components.WaldoStatusChip
import com.whereswaldo.android.ui.designsystem.components.WaldoStatusTone
import com.whereswaldo.android.ui.designsystem.components.WaldoTextField
import com.whereswaldo.android.ui.designsystem.components.WaldoTopBar
import java.time.Instant

/** The one owner action awaiting user confirmation (specs/003-android-client.md §12.2: "owner
 * controls behind confirm dialogs"). A single sealed type + one [AlertDialog] call site keeps the
 * five confirm flows (rename/extend/rotate/kick/delete) from needing five near-identical dialogs. */
private sealed class PendingGroupAction {
    data class Rename(val newName: String) : PendingGroupAction()
    data class Extend(val endsAtIso: String, val label: String) : PendingGroupAction()
    data object Rotate : PendingGroupAction()
    data class Kick(val userId: String, val displayName: String) : PendingGroupAction()
    data object Delete : PendingGroupAction()
}

/**
 * The A5 group-detail screen (001-api-contract.md §12.3-§12.5, §12.7-§12.9, specs/003-android-
 * client.md §12.2). Owner controls (rename, extend/end, rotate, kick, delete) run through
 * [AlertDialog] confirmations — one of the "un-migrated Material3 primitives" `WaldoTheme` already
 * themes correctly (specs/003 §4.3, same rationale as `HistoryScreen`'s `DatePickerDialog`
 * exception): a modal-overlay primitive with real platform behavior, not a styling primitive.
 *
 * On `GROUP_EXPIRED` ([GroupDetailUiState.Expired]) the screen shows a brief [Toast] — the same
 * visible notice [GroupMapScreen] gives on its identical `Expired` treatment (specs/003 §4.3's
 * documented exception; the `Expired` state's own message alone would render for at most one
 * composition frame before [onLeft] navigates away, effectively imperceptible) — and then bounces
 * back to the groups list via [onLeft] (specs/003 §12.2's "SHOULD bounce the user back to the
 * groups list with a notice"). After a successful leave/delete ([GroupDetailUiState.Content.left])
 * the same [onLeft] fires with no `Toast` — that's a deliberate, already-confirmed user action
 * (via an [AlertDialog] confirm or the plain "Leave group" button), not a surprise the user needs
 * to be told about.
 *
 * **A6 addition** (specs/007-public-join-links.md §1/§4, specs/003-android-client.md §12.3): the
 * share-sheet text now carries the canonical `https://{joinLinkHost}/g#{code}` link (007 §1: "the
 * https form is the canonical one for sharing and QR") instead of the `waldo://` form — the
 * `waldo://` scheme remains a fully supported deep link ([WaldoNavHost]), it's just no longer what
 * this screen puts in the share text (007 §1 assigns that role to the landing page's own "open the
 * app" affordance, W1 scope). A QR code of the same link renders below the share button, generated
 * entirely on-device via [GroupQrCodeGenerator] — no networked QR-image service is ever involved
 * (007 §4's hard requirement; see that object's doc for the full dependency justification).
 */
@Composable
fun GroupDetailRoute(
    viewModel: GroupDetailViewModel,
    joinLinkHost: String,
    modifier: Modifier = Modifier,
    onLeft: () -> Unit = {},
    onOpenMap: (groupId: String) -> Unit = {},
) {
    val state by viewModel.state.collectAsState()
    val context = LocalContext.current

    LaunchedEffect(state) {
        val expired = state as? GroupDetailUiState.Expired
        if (expired != null) {
            Toast.makeText(context, expired.message, Toast.LENGTH_SHORT).show()
            onLeft()
            return@LaunchedEffect
        }
        if ((state as? GroupDetailUiState.Content)?.left == true) {
            onLeft()
        }
    }

    GroupDetailScreen(
        state = state,
        joinLinkHost = joinLinkHost,
        onRename = viewModel::rename,
        onExtend = viewModel::updateEndsAt,
        onRotateCode = viewModel::rotateCode,
        onKickMember = viewModel::kickMember,
        onDeleteGroup = viewModel::deleteGroup,
        onLeaveGroup = viewModel::leaveGroup,
        onOpenMap = onOpenMap,
        onRetry = viewModel::reload,
        modifier = modifier,
    )
}

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun GroupDetailScreen(
    state: GroupDetailUiState,
    joinLinkHost: String,
    modifier: Modifier = Modifier,
    onRename: (String) -> Unit = {},
    onExtend: (String) -> Unit = {},
    onRotateCode: () -> Unit = {},
    onKickMember: (String) -> Unit = {},
    onDeleteGroup: () -> Unit = {},
    onLeaveGroup: () -> Unit = {},
    onOpenMap: (groupId: String) -> Unit = {},
    onRetry: () -> Unit = {},
) {
    val context = LocalContext.current
    var pendingAction by remember { mutableStateOf<PendingGroupAction?>(null) }
    var renameDraft by remember { mutableStateOf("") }

    Column(modifier = modifier.fillMaxSize()) {
        WaldoTopBar(title = if (state is GroupDetailUiState.Content) state.name else "Group")

        when (state) {
            is GroupDetailUiState.Loading -> WaldoLoadingState(message = "Loading group…")

            is GroupDetailUiState.Error -> WaldoErrorState(
                title = "Couldn't load this group",
                message = state.message,
                onRetry = onRetry,
            )

            // Transient — GroupDetailRoute's LaunchedEffect is about to navigate away.
            is GroupDetailUiState.Expired -> WaldoLoadingState(message = state.message)

            is GroupDetailUiState.Content -> {
                Column(
                    modifier = Modifier
                        .fillMaxWidth()
                        .padding(WaldoTheme.spacing.md),
                    verticalArrangement = Arrangement.spacedBy(WaldoTheme.spacing.sm),
                ) {
                    WaldoCard {
                        Column(verticalArrangement = Arrangement.spacedBy(WaldoTheme.spacing.xs)) {
                            Row(horizontalArrangement = Arrangement.spacedBy(WaldoTheme.spacing.sm)) {
                                WaldoStatusChip(label = state.state, tone = groupStateTone(state.state))
                                WaldoStatusChip(
                                    label = GroupCountdownFormatter.format(state.endsAt, Instant.now().toString()),
                                    tone = WaldoStatusTone.Neutral,
                                )
                            }
                            WaldoStatusChip(label = GroupPolicyCopy.forPolicy(state.expiryPolicy), tone = WaldoStatusTone.Neutral)
                            // 001 §12.10 — the group map only serves `active` groups (410
                            // GROUP_EXPIRED otherwise, 005 §2.3); hide the entry point rather than
                            // let the user land on an immediate bounce-back.
                            if (state.state == "active") {
                                WaldoButton(
                                    text = "View map",
                                    style = WaldoButtonStyle.Secondary,
                                    onClick = { onOpenMap(state.groupId) },
                                )
                            }
                        }
                    }

                    WaldoCard {
                        Column(verticalArrangement = Arrangement.spacedBy(WaldoTheme.spacing.xs)) {
                            if (state.code != null) {
                                val joinLink = remember(joinLinkHost, state.code) {
                                    GroupJoinLinkBuilder.buildHttpsLink(joinLinkHost, state.code)
                                }
                                WaldoStatusChip(label = "Code: ${state.code}", tone = WaldoStatusTone.Success)
                                WaldoButton(
                                    text = "Share code",
                                    style = WaldoButtonStyle.Secondary,
                                    onClick = {
                                        val shareIntent = Intent(Intent.ACTION_SEND).apply {
                                            type = "text/plain"
                                            putExtra(
                                                Intent.EXTRA_TEXT,
                                                "Join my \"${state.name}\" group on Where's waldo — " +
                                                    "code ${state.code}\n$joinLink",
                                            )
                                        }
                                        context.startActivity(Intent.createChooser(shareIntent, "Share group code"))
                                    },
                                )
                                // A6 (specs/007 §4): on-device QR of the exact share link above —
                                // no networked QR-image service (GroupQrCodeGenerator's doc has the
                                // full justification). Sized to fill the card's width rather than a
                                // hardcoded dp constant (specs/003 §4.1's token vocabulary has no
                                // token for "QR code edge length"; aspectRatio(1f) keeps it square
                                // at whatever width the layout gives it).
                                Image(
                                    bitmap = remember(joinLink) { GroupQrCodeGenerator.toBitmap(joinLink).asImageBitmap() },
                                    contentDescription = "QR code to join \"${state.name}\" — $joinLink",
                                    modifier = Modifier
                                        .fillMaxWidth()
                                        .aspectRatio(1f)
                                        .padding(top = WaldoTheme.spacing.sm),
                                    filterQuality = FilterQuality.None,
                                )
                            } else {
                                WaldoStatusChip(label = "Code no longer available", tone = WaldoStatusTone.Neutral)
                            }
                        }
                    }

                    state.mutationError?.let { WaldoStatusChip(label = it, tone = WaldoStatusTone.Danger) }

                    if (state.isOwner) {
                        WaldoSectionHeader(title = "Owner controls")
                        WaldoCard {
                            Column(verticalArrangement = Arrangement.spacedBy(WaldoTheme.spacing.sm)) {
                                WaldoTextField(value = renameDraft, onValueChange = { renameDraft = it }, label = "Rename group")
                                Row(horizontalArrangement = Arrangement.spacedBy(WaldoTheme.spacing.sm)) {
                                    WaldoButton(
                                        text = "Rename",
                                        enabled = !state.isMutating && renameDraft.isNotBlank(),
                                        style = WaldoButtonStyle.Secondary,
                                        onClick = { pendingAction = PendingGroupAction.Rename(renameDraft) },
                                    )
                                    WaldoButton(
                                        text = "Rotate code",
                                        enabled = !state.isMutating,
                                        style = WaldoButtonStyle.Secondary,
                                        onClick = { pendingAction = PendingGroupAction.Rotate },
                                    )
                                }
                                WaldoButton(
                                    text = "End group now",
                                    enabled = !state.isMutating,
                                    style = WaldoButtonStyle.Secondary,
                                    onClick = {
                                        // 001 §12.4: endsAt MUST be > now, and "endsAt <= now + 5min"
                                        // is the server's "end it now" convenience — 60s ahead is
                                        // safely inside that window and can't race negative against
                                        // the server's own clock (unlike sending exactly `now`).
                                        pendingAction = PendingGroupAction.Extend(
                                            endsAtIso = Instant.now().plusSeconds(60).toString(),
                                            label = "end this group now",
                                        )
                                    },
                                )
                                WaldoButton(
                                    text = "Delete group",
                                    enabled = !state.isMutating,
                                    style = WaldoButtonStyle.Secondary,
                                    onClick = { pendingAction = PendingGroupAction.Delete },
                                )
                            }
                        }
                    } else {
                        WaldoButton(
                            text = "Leave group",
                            enabled = !state.isMutating,
                            style = WaldoButtonStyle.Secondary,
                            onClick = onLeaveGroup,
                        )
                    }
                }

                val members = state.members
                if (members == null) {
                    WaldoStatusChip(
                        label = "Member list hidden while this group is ending",
                        tone = WaldoStatusTone.Neutral,
                        modifier = Modifier.padding(horizontal = WaldoTheme.spacing.md),
                    )
                } else {
                    WaldoSectionHeader(title = "Members", modifier = Modifier.padding(horizontal = WaldoTheme.spacing.md))
                    LazyColumn(
                        modifier = Modifier.padding(horizontal = WaldoTheme.spacing.md),
                        verticalArrangement = Arrangement.spacedBy(WaldoTheme.spacing.xs),
                    ) {
                        items(members, key = { it.userId }) { member ->
                            val kickAction: (@Composable RowScope.() -> Unit)? = if (state.isOwner && member.role != "owner") {
                                {
                                    WaldoButton(
                                        text = "Kick",
                                        style = WaldoButtonStyle.Secondary,
                                        onClick = { pendingAction = PendingGroupAction.Kick(member.userId, member.displayName) },
                                    )
                                }
                            } else {
                                null
                            }
                            WaldoListRow(
                                title = member.displayName,
                                subtitle = member.role,
                                trailing = kickAction,
                            )
                        }
                    }
                }
            }
        }
    }

    val action = pendingAction
    if (action != null) {
        GroupActionConfirmDialog(
            action = action,
            onConfirm = {
                when (action) {
                    is PendingGroupAction.Rename -> onRename(action.newName)
                    is PendingGroupAction.Extend -> onExtend(action.endsAtIso)
                    is PendingGroupAction.Rotate -> onRotateCode()
                    is PendingGroupAction.Kick -> onKickMember(action.userId)
                    is PendingGroupAction.Delete -> onDeleteGroup()
                }
                pendingAction = null
            },
            onDismiss = { pendingAction = null },
        )
    }
}

@Composable
private fun GroupActionConfirmDialog(
    action: PendingGroupAction,
    onConfirm: () -> Unit,
    onDismiss: () -> Unit,
) {
    val message = when (action) {
        is PendingGroupAction.Rename -> "Rename this group to \"${action.newName}\"?"
        is PendingGroupAction.Extend -> "Are you sure you want to ${action.label}?"
        is PendingGroupAction.Rotate -> "Rotate the join code? The old code will stop working immediately."
        is PendingGroupAction.Kick -> "Remove ${action.displayName} from this group?"
        is PendingGroupAction.Delete -> "Delete this group? Everything — members, positions, the join code — is removed immediately and cannot be undone."
    }
    AlertDialog(
        onDismissRequest = onDismiss,
        text = { Text(message) },
        confirmButton = { WaldoButton(text = "Confirm", onClick = onConfirm) },
        dismissButton = { WaldoButton(text = "Cancel", onClick = onDismiss, style = WaldoButtonStyle.Secondary) },
    )
}

private fun groupStateTone(state: String): WaldoStatusTone = when (state) {
    "active" -> WaldoStatusTone.Success
    "ended" -> WaldoStatusTone.Warning
    else -> WaldoStatusTone.Neutral
}

@Preview(name = "Group detail — light (owner)", showBackground = true)
@Composable
private fun GroupDetailScreenLightPreview() {
    WaldoTheme(darkTheme = false) {
        GroupDetailScreen(
            state = GroupDetailUiState.Content(
                groupId = "grp_9J2Kq7Lm3NpR5sTvWxYz",
                name = "Festival crew",
                endsAt = "2026-08-02T22:00:00Z",
                expiryPolicy = "delete",
                state = "active",
                role = "owner",
                memberCount = 2,
                code = "7F3K9QRZ",
                createdAt = "2026-07-21T10:00:00Z",
                members = listOf(
                    GroupMemberUi("u1", "Eric", "owner", "2026-07-21T10:00:00Z"),
                    GroupMemberUi("u9", "Noor", "member", "2026-07-21T10:05:00Z"),
                ),
            ),
            joinLinkHost = "CHANGE-ME.azurestaticapps.net",
        )
    }
}

@Preview(name = "Group detail — dark (member, roster hidden)", showBackground = true)
@Composable
private fun GroupDetailScreenDarkPreview() {
    WaldoTheme(darkTheme = true) {
        GroupDetailScreen(
            state = GroupDetailUiState.Content(
                groupId = "grp_9J2Kq7Lm3NpR5sTvWxYz",
                name = "Festival crew",
                endsAt = "2026-07-21T22:00:00Z",
                expiryPolicy = "grace",
                state = "ended",
                role = "member",
                memberCount = 2,
                code = null,
                createdAt = "2026-07-21T10:00:00Z",
                members = null,
            ),
            joinLinkHost = "CHANGE-ME.azurestaticapps.net",
        )
    }
}
